kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1522 @@
+        #pragma omp parallel for firstprivate(equation_size)
+        for (int index = 0; index < static_cast<int>(equation_size); ++index)
+            indices[index].reserve(40);
+
+        /// Definition of the eqautio id vector type
+        EquationIdVectorType ids(3, 0);
+        EquationIdVectorType second_ids(3, 0); // NOTE: Used only on the constraints to take into account the master dofs
+
+        #pragma omp parallel firstprivate(ids, second_ids)
+        {
+            // The process info
+            ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+            // We repeat the same declaration for each thead
+            std::vector<IndexSetType> temp_indexes(equation_size);
+
+            #pragma omp for
+            for (int index = 0; index < static_cast<int>(equation_size); ++index)
+                temp_indexes[index].reserve(30);
+
+            // Getting the size of the array of elements from the model
+            const int number_of_elements = static_cast<int>(rModelPart.Elements().size());
+
+            // Element initial iterator
+            const auto el_begin = rModelPart.ElementsBegin();
+
+            // We iterate over the elements
+            #pragma omp for schedule(guided, 512) nowait
+            for (int i_elem = 0; i_elem<number_of_elements; ++i_elem) {
+                auto it_elem = el_begin + i_elem;
+                pScheme->EquationId( *(it_elem.base()), ids, r_current_process_info);
+
+                for (auto& id_i : ids) {
+                    if (id_i < BaseType::mEquationSystemSize) {
+                        auto& row_indices = temp_indexes[id_i];
+                        for (auto& id_j : ids)
+                            if (id_j < BaseType::mEquationSystemSize)
+                                row_indices.insert(id_j);
+                    }
+                }
+            }
+
+            // Getting the size of the array of the conditions
+            const int number_of_conditions = static_cast<int>(rModelPart.Conditions().size());
+
+            // Condition initial iterator
+            const auto cond_begin = rModelPart.ConditionsBegin();
+
+            // We iterate over the conditions
+            #pragma omp for schedule(guided, 512) nowait
+            for (int i_cond = 0; i_cond<number_of_conditions; ++i_cond) {
+                auto it_cond = cond_begin + i_cond;
+                pScheme->Condition_EquationId( *(it_cond.base()), ids, r_current_process_info);
+                for (auto& id_i : ids) {
+                    if (id_i < BaseType::mEquationSystemSize) {
+                        auto& row_indices = temp_indexes[id_i];
+                        for (auto& id_j : ids)
+                            if (id_j < BaseType::mEquationSystemSize)
+                                row_indices.insert(id_j);
+                    }
+                }
+            }
+
+            // Getting the size of the array of the constraints
```

Comment:
A without modification contains all dofs that are not fixed (as a typical elimination B&S), this includes the constraint DoFs. The T contains in the row all the dofs from th previous A and in the column the dofs that are actually solved. Of course if your system has very few constraints this will be slower than a regular solve, the idea is to use it in system with many constraints. In any case the computation can be improved if the save the structure of the resulting matrix T'AT, then we will only fill the matrix, not compute the structure, which is more than a half of the multiplication operation.